pkg/scheduler/schedule_one.go

```diff
@@ -89,10 +90,43 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 	podsToActivate := framework.NewPodsToActivate()
 	state.Write(framework.PodsToActivateKey, podsToActivate)
 
-	schedulingCycleCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	scheduleResult, assumedPodInfo := sched.schedulingCycle(schedulingCycleCtx, state, fwk, podInfo, podsToActivate, start)
-	if scheduleResult.FeasibleNodes == 0 {
+	var scheduleResult ScheduleResult
+	var assumedPodInfo *framework.QueuedPodInfo
+
+	func() {
+		schedulingCycleCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		scheduleResult, assumedPodInfo, err = sched.schedulingCycle(schedulingCycleCtx, state, fwk, podInfo, start)
+		if err != nil {
+			switch scheduleResult.reason {
+			case v1.PodReasonUnschedulable:
+				metrics.PodUnschedulable(fwk.ProfileName(), metrics.SinceInSeconds(start))
+			case v1.PodReasonSchedulerError:
+				metrics.PodScheduleError(fwk.ProfileName(), metrics.SinceInSeconds(start))
+			}
+
+			if scheduleResult.unreserve {
+				// trigger un-reserve to clean up state associated with the reserved Pod
+				fwk.RunReservePluginsUnreserve(schedulingCycleCtx, state, assumedPodInfo.Pod, scheduleResult.SuggestedHost)
+				if forgetErr := sched.Cache.ForgetPod(assumedPodInfo.Pod); forgetErr != nil {
+					klog.ErrorS(forgetErr, "Scheduler cache ForgetPod failed")
+				}
+			}
+
+			sched.FailureHandler(schedulingCycleCtx, fwk, assumedPodInfo, err, scheduleResult.reason, scheduleResult.nominatingInfo)
+			return
+		}
+
+		// At the end of a successful scheduling cycle, pop and move up Pods if needed.
+		if len(podsToActivate.Map) != 0 {
+			sched.SchedulingQueue.Activate(podsToActivate.Map)
+			// Clear the entries after activation.
+			podsToActivate.Map = make(map[string]*v1.Pod)
+		}
+	}()
+
+	if err != nil {
 		return
 	}
```

Comment:
If we don't wrap the schedulingCycle in a function, we don't need this actually.

Reply:
I thought about this, but for readability I picked the currently implement.

Reply:
Another approach I thought about is `schedulingCycle := func() error {}`, then we call like this
```
if err := schedulingCycle(); err != nil {
    return
}
```

No big difference.